src/storage/mod.rs

```diff
@@ -2495,22 +2576,43 @@ impl<E: Engine, L: LockManager, F: KvFormat> Storage<E, L, F> {
         ttl: u64,
         cb: Callback<(Option<Value>, bool)>,
     ) -> Result<()> {
-        Self::check_api_version(
-            self.api_version,
-            ctx.api_version,
-            CommandKind::raw_compare_and_swap,
-            [&key],
-        )?;
-        let cf = Self::rawkv_cf(&cf, self.api_version)?;
+        const CMD: CommandKind = CommandKind::raw_compare_and_swap;
+        let api_version = self.api_version;
+        Self::check_api_version(api_version, ctx.api_version, CMD, [&key])?;
+        let cf = Self::rawkv_cf(&cf, api_version)?;
 
         if !F::IS_TTL_ENABLED && ttl != 0 {
             return Err(Error::from(ErrorInner::TtlNotEnabled));
         }
-
-        let key = F::encode_raw_key_owned(key, None);
-        let cmd =
-            RawCompareAndSwap::new(cf, key, previous_value, value, ttl, self.api_version, ctx);
-        self.sched_txn_command(cmd, cb)
+        let provider = self.causal_ts_provider.clone();
+        let sched = self.get_scheduler();
+        let concurrency_manager = self.get_concurrency_manager();
+        self.sched_raw_command(CMD, async move {
+            let key_guard = Self::get_raw_key_guard(&provider, concurrency_manager).await;
```

Comment:
We can use the key guard here to avoid the other locking (by latch) in `sched_raw_atomic_command`. Or leave a comment to do it later. The same to other atomic write requests.

Reply:
Use ts encoded key to get the lock guard to avoid lock twice. Maybe remove latch later, however much risk as it's on txn process, `TODO` is added.